File: Loenn/triggers/forceHiccupTrigger.lua
local ForceHiccupTrigger = {}

ForceHiccupTrigger.name = "KoseiHelper/ForceHiccupTrigger"
ForceHiccupTrigger.depth = 100

ForceHiccupTrigger.placements = {
	{
		name = "Force Hiccup Trigger",
		data = {
		onlyOnce = false,
		triggerMode = "OnEnter",
		interval = 0.5,
		flag = "cold"
		}
	}
}

function ForceHiccupTrigger.ignoredFields(entity)
	local ignored = {
	"_name",
	"_id",
	"interval",
	"flag"
	}
	local function doNotIgnore(value)
        for i = #ignored, 1, -1 do
            if ignored[i] == value then
                table.remove(ignored, i)
                return
            end
        end
    end
	if entity.triggerMode == "OnStay" then
		doNotIgnore("interval")
	end
	if entity.triggerMode == "OnFlagEnabled" then
		doNotIgnore("flag")
	end
	return ignored
end

ForceHiccupTrigger.fieldInformation = function (entity) return {
	triggerMode = {
		options = {
		"OnEnter",
		"OnLeave",
		"OnStay",
		"OnFlagEnabled",
		"OnCassetteBeat"
		},
		editable = false
	}
}
end

return ForceHiccupTrigger

File: Source/Triggers/ForceHiccupTrigger.cs
using Celeste.Mod.Entities;
using Microsoft.Xna.Framework;
using Monocle;

namespace Celeste.Mod.KoseiHelper.Triggers;


[CustomEntity("KoseiHelper/ForceHiccupTrigger")]
public class ForceHiccupTrigger : Trigger
{
    public bool onlyOnce;
    public TriggerMode triggerMode;
    private CassetteBlockManager cassetteManager;
    private int cassetteIndex;
    private bool canHiccup;
    public float interval;
    public string flag;
    private bool previousHasHiccuppedFromFlag, currentHasHiccuppedFromFlag;

    public ForceHiccupTrigger(EntityData data, Vector2 offset) : base(data, offset)
    {
        onlyOnce = data.Bool("onlyOnce", false);
        triggerMode = data.Enum("triggerMode", TriggerMode.OnEnter);
        interval = data.Float("interval", 0.5f);
        flag = data.Attr("flag", "cold");
    }

    public override void Awake(Scene scene)
    {
        base.Awake(scene);
        cassetteManager = scene.Tracker.GetEntity<CassetteBlockManager>();
    }

    public override void OnStay(Player player)
    {
        base.OnStay(player);
        Level level = SceneAs<Level>();
        if (player.Scene != null && triggerMode == TriggerMode.OnStay)
        {
            if (Scene.OnInterval(interval))
                player.HiccupJump();
        }
        currentHasHiccuppedFromFlag = level.Session.GetFlag(flag);
        if (player.Scene != null && triggerMode == TriggerMode.OnFlagEnabled)
        {
            if (currentHasHiccuppedFromFlag && !previousHasHiccuppedFromFlag)
            {
                player.HiccupJump();
            }
        }
        previousHasHiccuppedFromFlag = currentHasHiccuppedFromFlag;
    }

    public override void OnEnter(Player player)
    {
        base.OnEnter(player);
        if (player.Scene != null && triggerMode == TriggerMode.OnEnter)
        {
            player.HiccupJump();
            if (onlyOnce)
                RemoveSelf();
        }
    }

    public override void OnLeave(Player player)
    {
        base.OnLeave(player);
        if (player.Scene != null && triggerMode == TriggerMode.OnLeave)
            player.HiccupJump();
        if (onlyOnce)
            RemoveSelf();
    }

    public override void Update()
    {
        base.Update();
        Player player = Scene.Tracker.GetEntity<Player>();
        if (cassetteManager != null && triggerMode == TriggerMode.OnCassetteBeat && player != null)
        {
            if (cassetteIndex != cassetteManager.currentIndex)
            {
                cassetteIndex = cassetteManager.currentIndex;
                canHiccup = true;
            }
            if (canHiccup)
                player.HiccupJump();
        }
        canHiccup = false;
    }
}